File: pipeline.py
"""ZipBDR Research Pipeline — Main Entry Point.

Run with: python pipeline.py
Processes accounts from accounts.json one at a time through the full research pipeline.
"""

import asyncio
import json
import sys
from pathlib import Path

from dotenv import load_dotenv

load_dotenv()  # populate os.environ from .env before researchers/tools import

from orchestrator import process_account  # noqa: E402


async def main():
    """Process all accounts — one at a time."""
    accounts_file = Path("accounts.json")

    if not accounts_file.exists():
        print("ERROR: accounts.json not found. Create it with your account list.")
        print('Format: [{"name": "Acme Corp", "domain": "acmecorp.com", "ticker": "ACME"}]')
        sys.exit(1)

    with open(accounts_file) as f:
        accounts = json.load(f)

    print(f"\n{'='*60}")
    print(f"  LaszyResearch Pipeline — Processing {len(accounts)} accounts")
    print(f"{'='*60}\n")

    results = []
    failures = []

    for i, account in enumerate(accounts):
        print(f"[{i+1}/{len(accounts)}] Processing: {account['name']} ({account['domain']})")
        try:
            result = await process_account(account)
            score = result.get("score")
            if score is None:
                print(f"  → Scoring agent returned no score: {result.get('error', 'unknown')}")
                failures.append({"account": account["name"], "error": "no score returned"})
                continue
            print(f"  → Score: {score}/10 ({result.get('confidence', 'unknown')})")
            results.append(result)
        except Exception as e:
            print(f"  → FAILED: {e}")
            failures.append({"account": account["name"], "error": str(e)})
            continue

    # Summary
    print(f"\n{'='*60}")
    print(f"  COMPLETE: {len(results)} scored, {len(failures)} failed")
    if results:
        scored = [r for r in results if isinstance(r.get("score"), (int, float))]
        if scored:
            avg_score = sum(r["score"] for r in scored) / len(scored)
            print(f"  Average score: {avg_score:.1f}/10")
            top = sorted(scored, key=lambda r: r["score"], reverse=True)[:5]
            print(f"  Top accounts:")
            for r in top:
                print(f"    • {r['company']} — {r['score']}/10")
    print(f"{'='*60}\n")


if __name__ == "__main__":
    asyncio.run(main())

File: src/pipeline.ts
import { readFile } from "node:fs/promises";
import { resolve } from "node:path";

import { processAccount } from "./orchestrator.js";
import type { Account, ScoredAccount } from "./types.js";

async function loadAccounts(): Promise<Array<Partial<Account>>> {
  const accountsPath = resolve("accounts.json");
  try {
    const raw = await readFile(accountsPath, "utf8");
    const parsed = JSON.parse(raw) as unknown;
    if (!Array.isArray(parsed)) {
      throw new Error("accounts.json must contain an array of accounts");
    }
    return parsed as Array<Partial<Account>>;
  } catch (error) {
    if ((error as NodeJS.ErrnoException).code === "ENOENT") {
      console.error("ERROR: accounts.json not found. Create it with your account list.");
      console.error(
        'Format: [{"name": "Acme Corp", "domain": "acmecorp.com", "ticker": "ACME"}]'
      );
      process.exitCode = 1;
      return [];
    }
    throw error;
  }
}

export async function main(): Promise<void> {
  const accounts = await loadAccounts();
  if (accounts.length === 0 && process.exitCode) {
    return;
  }

  console.log(`\n${"=".repeat(60)}`);
  console.log(`  LaszyResearch Pipeline - Processing ${accounts.length} accounts`);
  console.log(`${"=".repeat(60)}\n`);

  const results: ScoredAccount[] = [];
  const failures: Array<{ account: string; error: string }> = [];

  for (const [index, account] of accounts.entries()) {
    const accountLabel = account.name ?? account.domain ?? `row ${index + 1}`;
    const domainLabel = account.domain ?? "unknown domain";
    console.log(
      `[${index + 1}/${accounts.length}] Processing: ${accountLabel} (${domainLabel})`
    );

    try {
      const result = await processAccount(account);
      console.log(`  -> Score: ${result.score}/100 (${result.confidence})`);
      results.push(result);
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      console.log(`  -> FAILED: ${message}`);
      failures.push({ account: accountLabel, error: message });
    }
  }

  console.log(`\n${"=".repeat(60)}`);
  console.log(`  COMPLETE: ${results.length} scored, ${failures.length} failed`);
  if (results.length > 0) {
    const avgScore =
      results.reduce((total, result) => total + result.score, 0) / results.length;
    console.log(`  Average score: ${avgScore.toFixed(1)}/100`);
    console.log("  Top accounts:");
    for (const result of [...results].sort((a, b) => b.score - a.score).slice(0, 5)) {
      console.log(`    - ${result.company} - ${result.score}/100`);
    }
  }
  console.log(`${"=".repeat(60)}\n`);
}

if (import.meta.url === `file://${process.argv[1]}`) {
  await main();
}
